File: Maximum Product Subarray.java
class Solution {    
    
    int maxProduct(int[] arr) {
    int n=arr.length;
    int sum=1;
    int ans=Integer.MIN_VALUE;
    for(int i=0;i<n;i++){
        int a=arr[i];
        sum*=a;
        ans=Math.max(ans,sum);
        if(sum==0)sum=1;
    }
    sum=1;
    for(int i=n-1;i>=0;i--){
    int a=arr[i];
    sum*=a;
    ans=Math.max(ans,sum);
    if(sum==0)sum=1;
    }   
    return ans;
    }
    
    
}

File: Maximum Product Subarray.cpp
class Solution {
  public:
    int maxProduct(vector<int> &arr) {
        // code here
        int n =arr.size();
        
        int ans = arr[0];
        int maxAns = ans, minAns = ans;
        
        for(int i=1; i<n; i++){
            if(arr[i] < 0)
                swap(maxAns, minAns);
                
            maxAns = max(arr[i], maxAns*arr[i]);
            minAns = min(arr[i], minAns*arr[i]);
            
            ans = max(ans, maxAns);
        }
        
        return ans;
    }
};
//GFG POTD solution for 08 February
